swarm/storage/resource.go

```diff
@@ -535,56 +576,47 @@ func (self *ResourceHandler) lookup(rsrc *resource, period uint32, version uint3
 	return nil, NewResourceError(ErrNotFound, "no updates found")
 }
 
-// load existing mutable resource into resource struct
-func (self *ResourceHandler) loadResource(nameHash common.Hash, name string, refresh bool) (*resource, error) {
-
-	if name == "" {
-		name = nameHash.Hex()
+// Retrieves a resource root chunk and creates/updates the index entry for it
+// with the resulting metadata
+func (self *ResourceHandler) LoadResource(key Key) (*resource, error) {
+	chunk, err := self.chunkStore.get(key, defaultRetrieveTimeout)
+	if err != nil {
+		return nil, err
 	}
 
-	// if the resource is not known to this session we must load it
-	// if refresh is set, we force load
-	rsrc := self.getResource(name)
-	if rsrc == nil || refresh {
-		rsrc = &resource{}
-		// make sure our name is safe to use
-		if !isSafeName(name) {
-			return nil, NewResourceError(ErrInvalidValue, fmt.Sprintf("Invalid name '%s'", name))
-		}
-		rsrc.name = &name
-		rsrc.nameHash = nameHash
-
-		// get the root info chunk and update the cached value
-		chunk, err := self.chunkStore.get(Key(rsrc.nameHash[:]), defaultRetrieveTimeout)
-		if err != nil {
-			return nil, err
-		}
+	// minimum sanity check for chunk data (an update chunk first two bytes is headerlength uint16, and cannot be 0)
+	// \TODO this is not enough to make sure the data isn't bogus. A normal content addressed chunk could still satisfy these criteria
+	if !bytes.Equal(chunk.SData[:2], []byte{0x0, 0x0}) {
+		return nil, NewResourceError(ErrCorruptData, fmt.Sprintf("Chunk is not a resource root chunk"))
+	} else if len(chunk.SData) <= indexSize {
+		return nil, NewResourceError(ErrNothingToReturn, fmt.Sprintf("Invalid chunk length %d, should be minimum %d", len(chunk.SData), indexSize+1))
+	}
 
-		// minimum sanity check for chunk data
-		if len(chunk.SData) != indexSize {
-			return nil, NewResourceError(ErrNothingToReturn, fmt.Sprintf("Invalid chunk length %d, should be %d", len(chunk.SData), indexSize))
-		}
-		rsrc.startBlock = binary.LittleEndian.Uint64(chunk.SData[2:10])
-		rsrc.frequency = binary.LittleEndian.Uint64(chunk.SData[10:])
-	} else {
-		rsrc.name = self.resources[name].name
-		rsrc.nameHash = self.resources[name].nameHash
-		rsrc.startBlock = self.resources[name].startBlock
-		rsrc.frequency = self.resources[name].frequency
+	// create the index entry
+	rsrc := &resource{
+		startBlock: binary.LittleEndian.Uint64(chunk.SData[2:10]),
```

Comment:
dude... AS400 style data structures. why can't this be a JSON? you already have it in a struct - just add json marshalling annotations and marshal/unmarshal

Reply:
marshal/unmarshal.

Reply:
Encoding resource type may not need the flexibility of json. Low level binary encoding should be good as I assume representation will probably have just minor changes in the future. In that case, it would be helpful to implement encoding BinaryMarshaler and encoding.BinaryUnmarshaler to provide a transparent way for serialization. Also, this approach will make handling of future serialization changes easier.

Reply:
If for serialization only I also think that a binary encoding is fine imo. And what @janos proposes sounds like a clean interface to do that.